src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -21,36 +21,67 @@ internal static class CompletionHelpers
         /// <param name="wordToComplete">The word to complete.</param>
         /// <param name="possibleCompletionValues">The possible completion values to iterate.</param>
         /// <param name="toolTipMapping">The optional tool tip mapping delegate.</param>
+        /// <param name="listItemTextMapping">The optional list item text mapping delegate.</param>
         /// <param name="resultType">The optional completion result type. Default is Text.</param>
-        /// <returns></returns>
+        /// <returns>List of matching completion results.</returns>
         internal static IEnumerable<CompletionResult> GetMatchingResults(
             string wordToComplete,
             IEnumerable<string> possibleCompletionValues,
             Func<string, string> toolTipMapping = null,
+            Func<string, string> listItemTextMapping = null,
             CompletionResultType resultType = CompletionResultType.Text)
         {
             string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
-            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
 
             foreach (string value in possibleCompletionValues)
             {
-                if (pattern.IsMatch(value))
+                if (IsMatch(value, wordToComplete))
                 {
                     string completionText = QuoteCompletionText(value, quote);
+                    string toolTip = toolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = listItemTextMapping?.Invoke(value) ?? value;
```

Comment:
We should do the both mappings in one method. If this requires many changes in other completers it is better to implement this in follow PR.

Reply:
Agree. I am actually thinking of just changing the type from `Func<string, string>` to `Dictionary<string, (string Tooltip, string ListItemText)>`.  I am not sure `Func` is the best type to use here.

Reply:
It could be `Func<string, (string Tooltip, string ListItemText)>`. It's not critical right now. We can postpone the decision when we choose how best to extend the interface.

Reply:
Yep sounds good. Will postpone this. I'd like to just get IsMatch and ListItemTextMapping in interface in next PR.

Reply:
> We should do the both mappings in one method.

`+` for flexibility we could consider cascading as for possible values  `method` => `static list`.